mockrxandroidble/README.md

```diff
@@ -38,7 +38,16 @@ RxBleClient rxBleClientMock = new RxBleClientMock.Builder()
                             .addDescriptor(descriptorUUID, descriptorData) // <-- adding descriptor mocks
                             .build() // to the characteristic, there can be multiple of them
                     ).build()
-            ).build()
+            ).characteristicReadCallback(characteristicUUID, (characteristic) -> {
+                return Single.just(characteristicValue);
+            })
+            .characteristicWriteCallback(characteristicUUID, (characteristic, bytes) -> {
+                if(!writeData(characteristic, bytes)) {
+                    throw new BleGattCharacteristicException(null, characteristic, 0x80, BleGattOperationType.CHARACTERISTIC_WRITE);
+                    
+                }
```

Comment:
I am thinking if a better idea would be to have something along the lines:
```
RxBleExceptionMock.failOperation(status)
```
or
```
RxBleExceptionMock.failConnection(status, afterTimeout /* 0-30 seconds */)
```
And map it accordingly without user's input? This would need more work on the library side but the user would not need to know insides of the lib and have a separate API.

Or we could have a WriteResponse interface as an argument in `characteristicWriteCallback`
```
interface WriteResponse {
  Completable success();
  Completable failure(int status);
  Completable disconnect(int status);
}
```
And a similar for `ReadResponse`

WDYT? I have no hard opinions as I hardly use MockClient — just some loose thoughts.

Reply:
```
RxBleExceptionMock.failOperation(status)
```
Would this return a Completable error or throw?

> interface WriteResponse {
>   Completable success();
>   Completable failure(int status);
>   Completable disconnect(int status);
> }

This would be a parameter to the callback function?

Reply:
The first idea was that it could throw. But after some thought I would say it should return a Completable so the user could postpone it as they wish.
Additionally it should be an instance call, not static as exceptions with different parameters should be emitted for different operations.

Reply:
>Additionally it should be an instance call, not static as exceptions with different parameters should be emitted for different operations.

Could you explain further, I'm not clear on the reasoning here?

Reply:
Ok got it

Reply:
These are just some thoughts on the API side for you to think on. I will be unavailable next week (holiday finally!) so don't be discouraged if I will not respond promptly.

Reply:
I started implementation but I'm not sure which would work best. Either utility classes to create the `Single`s and `Completable`s, or a parameter on the callback with an interface like:
```
interface WriteResponse {
void success();
void failure(int status);
void disconnect(int status);
}
```
So the callback would look like this:
```
(characteristic, data, response) -> {
    if(writeData(characteristic, data)) {
        response.success();
    } else {
        response.failure(0x80);
    }
}
```
The problem with the former is that the utility methods would need to accept the characteristic (for failure) or device (for disconnect) as parameters, requiring the user to do more. The problem with the latter is it is less 'reactive' and requires them to maintain a reference to the "response" object. Any thoughts?

Reply:
Utility class is harder to discover (as it needs knowledge a priori) when compared with a "response" object that is passed as an argument.
I am not getting the 'less reactive' point — could you elaborate?
For me reactiveness of the latter approach is just fine — the response object is a new instance for each request and can be passed down the chain of operators as needed.

Reply:
Ok sounds like a response parameter is the way to go then. By less reactive I just meant it’s not returning an Observable. But I’m pretty new to ReactiveX so you know better than me!